File: C++/Hard/Accepted/k-th-smallest-in-lexicographical-order_6.cpp
class Solution {
public:
    int findKthNumber(int n, int k) {
        
        int curr = 1;
        k = k - 1;
        while (k > 0) {
            long long steps = getStep(n, curr, curr + 1);
            if (steps <= k) {
                curr += 1;
                k -= steps;
            } else {
                curr *= 10;
                k -= 1;
            }
        }
        return curr;
    }
    long long getStep(int n, long long n1, long long n2) {
        long long step = 0;
        while (n1 <= n) {
            step += min((long long)n + 1, n2) - n1;
            n1 *= 10;
            n2 *= 10;
        }
        return step;
    }
};

File: Python3/Hard/WrongAnswer/k-th-smallest-in-lexicographical-order_6.py
class Solution:
    def findKthNumber(self, n: int, k: int) -> int:
        
        def count(prefix, n):
            c = 0
            while prefix <= n:
                c += min(n + 1, prefix + 10) - prefix
                prefix *= 10
            return c
        prefix = 1
        k -= 1
        while k > 0:
            c = count(prefix, n)
            if c <= k:
                prefix += 1
                k -= c
            else:
                prefix *= 10
                k -= 1
        return prefix